Reworked TMHM into expandable list format

gruxor opened the pull request:
## Description
Completely removed the need for bit arrays, instead using the format of `gLevelUpLearnsets` as a guide to rework TMs/HMs entirely. TMHM lists use `MOVE_UNAVAILABLE` as a footer to allow proper looping, since dynamic arrays can't have their sizes determined at compile time. `gLevelUpLearnsets` uses an identical method through `LEVEL_UP_END` which shares the same value (max size u16, i.e. 0xFFFF). Optionally, a new preprocessor define could be used instead for the footer to prevent issues with any future integer type modifications.

- Should be much friendlier for new faces and less of a headache for old faces, with virtually no difference in performance/memory that I'm aware of. 
- Should cover all use cases - functions that the data was/is fed into have all been accounted for across the entire project (there weren't many)
- Haven't yet touched the tutor movelist (which uses an almost identical bit array system) but I imagine someone wanting to expand that list is much more rare. Could be easily done in the future if necessary, or just for consistency/ease of use
- Movelists may have minor errors since I regex'd the old list during conversion, but I'm pretty sure the compiler already caught most if not all of those during the testing process
- Support for PoryMoves should be a fairly trivial modification since the app already supports leveled lists
- Probably conflicts with #2232 (sorry Stephen) as it's an alternative solution to the same problem, but without the 128-bit limitations and with added user-friendliness

## **Discord contact info**
Crater#7777

StephenLynx commented:
You skipped the use of P_NEW_POKEMON to enable some entries?

gruxor commented:
> You skipped the use of P_NEW_POKEMON to enable some entries?

Checks for `P_NEW_POKEMON` still exist in tmhm_learnset_pointers.h, which determines if the arrays are actually used. They're statically defined then added to `gTMHMLearnsets` and can't really be used without it (and it'd defeat the purpose to access them directly anyways). 

Preprocessor if statements could be introduced (re-introduced?) to tmhm_learnets.h, but I doubt it really makes much of a difference in the end.

AsparagusEduardo commented:
`P_NEW_POKEMON` is meant to save space, and therefore, any data related to the new mon shouldn't be compiled into the rom if it's not gonna be used.

gruxor commented:
Fixed.

StephenLynx commented:
Speaking of space, I think it's important to keep in mind this will at least double the space used for this, maybe even triple.

AsparagusEduardo reviewed (changes requested):
2 additional things:
- Spaces instead of tabs in `tmhm_learnsets.h`

- Something that I've been thinking is that in vanilla, TMs/HMs and Tutor Moves don't overlap in terms of moves that can be taught. And even if they did, I doubt that they would be able to be learned with one method and not the other. My thought is that we could potentially unify both TM and Tutor tables into a single list, seeing as not only `CanMonLearnTMTutor` is already used by both, but it would further simplify modifying both Tutors and TMs (eg: no need for `TUTOR_MOVE_*` and stuff).

gruxor commented on src/pokemon.c:
```diff
@@ -1890,17 +1890,18 @@ const s8 gNatureStatTable[NUM_NATURES][NUM_NATURE_STATS] =
 #include "data/pokemon/experience_tables.h"
 #include "data/pokemon/base_stats.h"
 #include "data/pokemon/level_up_learnsets.h"
+#include "data/pokemon/tmhm_learnsets.h"
```
The includes didn't seem sorted to me in any way, and I figured it'd make sense if the learnsets were near each other. I was going to include the planned tutor movelist under those two as well. 

(that and I think I ended up fiddling around with a few new files that in the end got merged and renamed to tmhm_learnsets.h - the position in the define list may have been a casualty of that process, not sure).

Either way, no consequence - can be reverted if you think it's best.

gruxor commented:
Updated the tabs to spaces.

> Something that I've been thinking is that in vanilla, TMs/HMs and Tutor Moves don't overlap in terms of moves that can be taught. And even if they did, I doubt that they would be able to be learned with one method and not the other. My thought is that we could potentially unify both TM and Tutor tables into a single list, seeing as not only CanMonLearnTMTutor is already used by both, but it would further simplify modifying both Tutors and TMs (eg: no need for TUTOR_MOVE_* and stuff).

I can see that working for _most_ use cases, but not all. For example, if you have one-use TMs (or TRs, if those are implemented) then you could give the player one for a fairly powerful move early on. Then later, maybe you want to grant access to a tutor that can teach it infinitely in exchange for a Star Piece or something. Can't imagine stuff like that would be super common, but it isn't out of the question.

I think it'd definitely still be possible to trim down some of the code if I set up the tutor movelists in the same way, though. I can give it a shot if it'd be better to merge it alongside this change rather than via a separate PR.

AsparagusEduardo commented:
> I can see that working for _most_ use cases, but not all. For example, if you have one-use TMs (or TRs, if those are implemented) then you could give the player one for a fairly powerful move early on. Then later, maybe you want to grant access to a tutor that can teach it infinitely in exchange for a Star Piece or something. Can't imagine stuff like that would be super common, but it isn't out of the question.

I don't how that couldn't work with what I proposed. If the mon is able to learn the move with a single-use TM, it should also be able to learn that move from that Star Piece Tutor if both instances are read from the same move compatibility table.

gruxor commented:
> I don't how that couldn't work with what I proposed. If the mon is able to learn the move with a single-use TM, it should also be able to learn that move from that Star Piece Tutor if both instances are read from the same move compatibility table.

That's true, I don't know why I was thinking the method would matter. The thing in question would be "should a species be able to learn a certain move from a TM and not a Tutor, or vice versa?" and I really can't think of any examples where anyone would want that.

I'll see what I can do, but I'll probably have to set up a script of some sort this time since I'll need to merge the arrays.

AsparagusEduardo commented:
> I'll see what I can do, but I'll probably have to set up a script of some sort this time since I'll need to merge the arrays.

Either that, or we could wait until I add beta PoryMoves support for this :eyes:. Don't know how long that would take though, I tend to get distracted a lot lol

gruxor commented:
Massive changes with [ed02c4b](https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/ed02c4b425028a4fe2cf622653f352ec10914736). Boom.

- Merged tutor and TMHM learnlists into one
- Removed incredibly stupid arg requirements for the "can this pokemon learn this move" functions - now take in a move ID instead of needing to pass through a dummy item ID or dummy tutor ID based on the usage
- Removed gTutorMoves entirely, tutors now use move IDs directly rather than an abstracted list
- Removed more than a handful redundant carry-overs from the leveled lists. Lists existed for forms that have no differences, such as the Vivillion colors and many others. These were already not being used, because the pointer lists never assigned anything to them, instead using their base form. I'll probably submit a separate PR to clean up the leveled lists in the same way since that's the source of the redundant lists.

DizzyEggg commented on src/pokemon.c:
```diff
@@ -7243,40 +7244,43 @@ bool8 TryIncrementMonLevel(struct Pokemon *mon)
-u32 CanMonLearnTMHM(struct Pokemon *mon, u8 tm)
+u32 CanMonLearnTaughtMove(struct Pokemon *mon, u16 move)
 {
     u16 species = GetMonData(mon, MON_DATA_SPECIES2, 0);
```
Wouldn't it make sense to make it so that CanMonLearnTaughtMove calls CanSpeciesLearnTaughtMove instead of duplicating code two times?

gruxor replied:
Definitely. Left it as-is to avoid additional opinionated changes, but I'm more than willing to make them.

I was considering just removing CanMonLearnTaughtMove entirely - the only thing it does differently is grabs the species from the mon via `GetMonData(mon, MON_DATA_SPECIES2, 0);` and that code can be added to the only 2 sections that use the function (CanMonLearnMove in party_menu.c and BuildEggMoveset in daycare.c). Not really a big difference either way - but imo it's a better practice to do data conversion like that in-context _before_ passing the value into a function instead of having multiple functions that do conversion in different ways. GetMonData is already widely used around the codebase to retrieve MON_DATA_SPECIES2 (93 results for uses with that enum member specifically) so I figure it'd be consistent.

Thoughts?

DizzyEggg replied:
Sure, go ahead

gruxor replied:
Changed in https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/b8a3cc284650eebe2357986b07ed14fe1f94a5c2.

AsparagusEduardo commented on src/field_specials.c:
```diff
@@ -3130,7 +3130,7 @@ void GetBattleFrontierTutorMoveIndex(void)
         do
         {
-            if (gTutorMoves[i] == sBattleFrontier_TutorMoves2[moveIndex])
+            if (i == sBattleFrontier_TutorMoves2[moveIndex])
```
`GetBattleFrontierTutorMoveIndex` needs to be properly adjusted for this change.

gruxor replied:
Fixed in [54ee960](https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/54ee960970c31a49ebd83dfb2d396c67b2e6abc1) and found a few other oversights.

AsparagusEduardo reviewed (changes requested):
Awesome, this is great to see!

Though, there are some things that I feel should be changed name-wise:
- `gTeachLearnsets` ->`gTMTutorLearnsets`
- `CanMonLearnTaughtMove` -> `CanMonLearnTMTutorMove`
- `CanSpeciesLearnTaughtMove` -> `CanSpeciesLearnTMTutorMove`
- `CanMonLearnMove` -> `CanMonLearnTMTutor` 

- `teach_learnset_pointers.h` ->`tm_tutor_learnset_pointers.h` 
- `teach_learnsets.h` -> `tm_tutor_learnsets.h` 
- `sSPECIESTeachLearnset` -> `sSPECIESTMTutorLearnset`

gruxor commented:
> ... Though, there are some things that I feel should be changed name-wise: ...

Discussed on Discord, updated in https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/f8fb4e8cf247f0ad4b164686bbf42697d7908eb1.

gruxor commented:
Rushed the initial changes to Battle Frontier and assumed some variables meant something they didn't actually. Now, having fully deciphered the spaghetti, the tutors should actually be fixed in https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/9f16d1dd124e065ff45b7cb42cccbaa9d519c70f. 

Gotta love multi-use unnamed variables.

AsparagusEduardo commented on src/apprentice.c:
```diff
@@ -338,7 +338,7 @@ static u16 GetRandomAlternateMove(u8 monId)
     bool32 needTMs = FALSE;
     u16 moveId = MOVE_NONE;
-    bool32 shouldUseMove;
+    bool8 shouldUseMove;
```
This can stay as `bool32`. Besides that looks good.

gruxor replied:
Restored in https://github.com/rh-hideout/pokeemerald-expansion/pull/2233/commits/1c735e1abe0c872e661d68066c62cf2a0d4ad05f.
